### ISSUE TITLE
WriteOperationCompletionCallback::OperationCompleted() holds lock for a long while when Executor::UpdateIndexes() is called

### ISSUE THREAD

**Issue by robertpang** [YugaByte/yugabyte-db#581]
This can happen when updating a table with secondary index:

```
Stack found (1): #0  syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38
#2  0x00007f7d17fc0070 in _M_load_and_test_until () at /home/centos/.linuxbrew-yb-build/Cellar/gcc/5.4.0/include/c++/5.4.0/bits/atomic_futex.h:104
#3  _M_load_and_test () at /home/centos/.linuxbrew-yb-build/Cellar/gcc/5.4.0/include/c++/5.4.0/bits/atomic_futex.h:122
#4  _M_load_when_equal () at /home/centos/.linuxbrew-yb-build/Cellar/gcc/5.4.0/include/c++/5.4.0/bits/atomic_futex.h:162
#5  wait () at /home/centos/.linuxbrew-yb-build/Cellar/gcc/5.4.0/include/c++/5.4.0/future:322
#6  _M_get_result () at /home/centos/.linuxbrew-yb-build/Cellar/gcc/5.4.0/include/c++/5.4.0/future:681
#7  get () at /home/centos/.linuxbrew-yb-build/Cellar/gcc/5.4.0/include/c++/5.4.0/future:760
#8  yb::ql::ExecContext::PrepareChildTransaction () at ../../src/yb/yql/cql/ql/exec/exec_context.cc:73
#9  0x00007f7d17fc7237 in yb::ql::Executor::UpdateIndexes () at ../../src/yb/yql/cql/ql/exec/executor.cc:1724
#10 0x00007f7d17fc5599 in yb::ql::Executor::AddOperation () at ../../src/yb/yql/cql/ql/exec/executor.cc:1831
#11 0x00007f7d17fc5888 in yb::ql::Executor::ExecPTNode () at ../../src/yb/yql/cql/ql/exec/executor.cc:990
#12 0x00007f7d17fca6d1 in yb::ql::Executor::ExecTreeNode () at ../../src/yb/yql/cql/ql/exec/executor.cc:211
#13 0x00007f7d17fcc995 in yb::ql::Executor::ProcessAsyncResults () at ../../src/yb/yql/cql/ql/exec/executor.cc:1375
#14 0x00007f7d17fccefa in yb::ql::Executor::FlushAsyncDone () at ../../src/yb/yql/cql/ql/exec/executor.cc:1319
#15 0x00007f7d17fcd42a in operator() () at ../../src/yb/yql/cql/ql/exec/executor.cc:1244
#16 boost::detail::function::void_function_obj_invoker1<yb::ql::Executor::FlushAsync()::<lambda(const yb::Status&)>, void, const yb::Status&>::invoke(boost::detail::function::function_buffer &, const yb::Status &) () at ../../thirdparty/installed/uninstrumented/include/boost/function/function_template.hpp:159
#17 0x00007f7d134ac767 in operator() () at ../../thirdparty/installed/uninstrumented/include/boost/function/function_template.hpp:760
#18 operator() () at ../../src/yb/client/batcher.cc:206
#19 std::_Function_handler<void(), yb::client::internal::Batcher::RunCallback(const yb::Status&)::<lambda()> >::_M_invoke(const std::_Any_data &) () at /home/centos/.linuxbrew-yb-build/Cellar/gcc/5.4.0/include/c++/5.4.0/functional:1871
#20 0x00007f7d134a9b36 in operator() () at /home/centos/.linuxbrew-yb-build/Cellar/gcc/5.4.0/include/c++/5.4.0/functional:2267
#21 Run () at ../../src/yb/util/threadpool.h:479
#22 yb::client::internal::Batcher::RunCallback () at ../../src/yb/client/batcher.cc:208
#23 0x00007f7d134ab21c in yb::client::internal::Batcher::CheckForFinishedFlush () at ../../src/yb/client/batcher.cc:201
#24 0x00007f7d134a1042 in yb::client::internal::AsyncRpc::Finished () at ../../src/yb/client/async_rpc.cc:158
#25 0x00007f7d1098b57a in operator() () at /home/centos/.linuxbrew-yb-build/Cellar/gcc/5.4.0/include/c++/5.4.0/functional:2267
#26 yb::rpc::OutboundCall::CallCallback () at ../../src/yb/rpc/outbound_call.cc:326
#27 0x00007f7d1098b901 in yb::rpc::OutboundCall::SetFinished () at ../../src/yb/rpc/outbound_call.cc:407
#28 0x00007f7d10993a88 in yb::rpc::LocalYBInboundCall::Respond () at ../../src/yb/rpc/local_call.cc:88
#29 0x00007f7d109d08f9 in yb::rpc::YBInboundCall::RespondSuccess () at ../../src/yb/rpc/yb_rpc.cc:329
#30 0x00007f7d109b2ffd in yb::rpc::RpcContext::RespondSuccess () at ../../src/yb/rpc/rpc_context.cc:138
#31 0x00007f7d15b1e01c in yb::tserver::WriteOperationCompletionCallback::OperationCompleted () at ../../src/yb/tserver/tablet_service.cc:320
#32 0x00007f7d157f3b0a in yb::tablet::OperationDriver::Finalize () at ../../src/yb/tablet/operations/operation_driver.cc:454
#33 0x00007f7d157f3ce0 in yb::tablet::OperationDriver::ApplyTask () at ../../src/yb/tablet/operations/operation_driver.cc:443
#34 0x00007f7d157f4069 in yb::tablet::OperationDriver::ApplyOperation () at ../../src/yb/tablet/operations/operation_driver.cc:418
#35 0x00007f7d157f4b42 in yb::tablet::OperationDriver::ReplicationFinished () at ../../src/yb/tablet/operations/operation_driver.cc:373
#36 0x00007f7d154eba18 in yb::consensus::ReplicaState::ApplyPendingOperationsUnlocked () at ../../src/yb/consensus/replica_state.cc:808
#37 0x00007f7d154ec383 in yb::consensus::ReplicaState::AdvanceCommittedIndexUnlocked () at ../../src/yb/consensus/replica_state.cc:746
#38 0x00007f7d154d5090 in yb::consensus::RaftConsensus::EarlyCommitUnlocked () at ../../src/yb/consensus/raft_consensus.cc:1599
#39 0x00007f7d154d7f41 in yb::consensus::RaftConsensus::UpdateReplica () at ../../src/yb/consensus/raft_consensus.cc:1537
#40 0x00007f7d154d8c42 in yb::consensus::RaftConsensus::Update () at ../../src/yb/consensus/raft_consensus.cc:1126
#41 0x00007f7d15b1a710 in yb::tserver::ConsensusServiceImpl::UpdateConsensus () at ../../src/yb/tserver/tablet_service.cc:1213
#42 0x00007f7d12d34c97 in yb::consensus::ConsensusServiceIf::Handle () at src/yb/consensus/consensus.service.cc:99
#43 0x00007f7d109c754a in yb::rpc::ServicePoolImpl::Handle () at ../../src/yb/rpc/service_pool.cc:214
#44 0x00007f7d109c667a in Run () at ../../src/yb/rpc/service_pool.cc:252
#45 yb::rpc::TasksPool<yb::rpc::(anonymous namespace)::InboundCallTask>::WrappedTask::Run () at ../../src/yb/rpc/tasks_pool.h:70
#46 0x00007f7d109cd9f9 in yb::rpc::(anonymous namespace)::Worker::Execute () at ../../src/yb/rpc/thread_pool.cc:98
#47 0x00007f7d0e9b0d22 in operator() () at /home/centos/.linuxbrew-yb-build/Cellar/gcc/5.4.0/include/c++/5.4.0/functional:2267
#48 yb::Thread::SuperviseThread () at ../../src/yb/util/thread.cc:603
#49 0x00007f7d09ed7694 in start_thread () at pthread_create.c:333
```

**Comment by robertpang** [comment]
Reorganizing the index update should avoid lock contention and give higher throughput.

**Comment by robertpang** [comment]
Addressed in commit [04eabfc](https://github.com/YugaByte/yugabyte-db/commit/04eabfcd753443b63a630badc65abb947df6eb67).
